src/MLIAP/mliap_descriptor_so3.cpp

```diff
@@ -0,0 +1,298 @@
+/* ----------------------------------------------------------------------
+   LAMMPS - Large-scale Atomic/Molecular Massively Parallel Simulator
+   https://lammps.sandia.gov/, Sandia National Laboratories
+   Steve Plimpton, sjplimp@sandia.gov
+
+   Copyright (2003) Sandia Corporation.  Under the terms of Contract
+   DE-AC04-94AL85000 with Sandia Corporation, the U.S. Government retains
+   certain rights in this software.  This software is distributed under
+   the GNU General Public License.
+
+   See the README file in the top-level LAMMPS directory.
+------------------------------------------------------------------------- */
+
+#include "mliap_descriptor_so3.h"
+#include "atom.h"
+#include "comm.h"
+#include "error.h"
+#include "memory.h"
+#include "mliap_data.h"
+#include "pair_mliap.h"
+#include "mliap_so3.h"
+
+#include <cmath>
+#include <cstring>
+
+using namespace LAMMPS_NS;
+
+#define MAXLINE 1024
+#define MAXWORD 3
+
+
+MLIAPDescriptorSO3::MLIAPDescriptorSO3(LAMMPS *lmp,
+                                       char *paramfilename):
+  MLIAPDescriptor(lmp)
+{
+  nelements = 0;
+  elements = nullptr;
+  radelem = nullptr;
+  wjelem = nullptr;
+  so3ptr = nullptr;
+  read_paramfile(paramfilename);
+
+  so3ptr = new MLIAP_SO3(lmp,rcutfac,lmax,nmax,alpha);
+
+  ndescriptors = so3ptr->ncoeff;
+}
+
+MLIAPDescriptorSO3::~MLIAPDescriptorSO3()
+{
+  delete so3ptr;
+}
+
+void MLIAPDescriptorSO3::read_paramfile(char *paramfilename)
+{
+  int rcutfacflag = 0;
+  int nelementsflag = 0;
+  int elementsflag = 0;
+  int radelemflag = 0;
+  int wjelemflag = 0;
+  int nmaxflag=0;
+  int lmaxflag=0;
+  int alphaflag=0;
+
+  // set defaults for optional keywords
+
+  rfac0 = 0.99363;
+  rmin0 = 0.0;
+
+  // open SO3 parameter file on proc 0
+
+  FILE *fpparam;
+  if (comm->me == 0) {
+    fpparam = utils::open_potential(paramfilename,lmp,nullptr);
+    if (fpparam == nullptr)
+      error->one(FLERR,fmt::format
+        ("Cannot open SO3 parameter file {}: {}",
+          paramfilename, utils::getsyserror()));
+  }
+
+  char line[MAXLINE],*ptr;
+  int eof = 0;
+  int n,nwords;
+
+  while (1) {
+    if (comm->me == 0) {
+      ptr = fgets(line,MAXLINE,fpparam);
+      if (ptr == nullptr) {
+        eof = 1;
+        fclose(fpparam);
+      } else n = strlen(line) + 1;
+    }
+
+    MPI_Bcast(&eof,1,MPI_INT,0,world);
+    if (eof) break;
+    MPI_Bcast(&n,1,MPI_INT,0,world);
+    MPI_Bcast(line,n,MPI_CHAR,0,world);
+
+    // strip comment, skip line if blank
+
+    if ((ptr = strchr(line,'#'))) *ptr = '\0';
+    nwords = utils::count_words(line);
+    if (nwords == 0) continue;
+
+    // words = ptrs to all words in line
+    // strip single and double quotes from words
+
+    char* keywd = strtok(line,"' \t\n\r\f");
+    char* keyval = strtok(nullptr,"' \t\n\r\f");
+
+    if (comm->me == 0) {
+      utils::logmesg(lmp, fmt::format("SO3 keyword {} {} \n",
```

Comment:
Also `utils::logmesg()` does accept [a format string and a variable number of arguments now](https://docs.lammps.org/Developer_utils.html#_CPPv4I0DpEN9LAMMPS_NS5utils7logmesgEvP6LAMMPSRK1SDpRR4Args)